python/custom_field_definitions/add_custom_field.py

```diff
@@ -0,0 +1,183 @@
+import json
+import requests
+
+# Dump JSON is a readable format.
+def print_json(json_obj):
+    print(json.dumps(json_obj, sort_keys=True, indent=2))
+
+def print_http_error(error_string, http_status_code, resp_json):
+    print(f"{error_string} {http_status_code}: {resp_json['message']}, {resp_json['error']}")
+    return
+
+# Create a dictionary of custom field defintions keyed by custom field defintion name.
+def get_custom_field_definitions(base_url, headers):
+    # Dictionary of custom field definitions.
+    custom_field_defs = {}
+
+    list_custom_field_defs_url = f"{base_url}custom_field_definitions"
+
+    # List the the custom fields.  Only the first 100 with the code.
+    response = requests.get(list_custom_field_defs_url, headers=headers)
+    resp_json = response.json()
+    if response.status_code != 200:
+        print_http_error("List Risk Meters Error: ", response.status_code, resp_json)
+        sys.exit(1)
+
+    custom_field_defs_resp = resp_json['custom_field_definitions']
+
+    # Create a custom field definition record from the API response.
+    for cdf in custom_field_defs_resp:
+        custom_field_def = {}
+        custom_field_def['name'] = cdf['name']
+        custom_field_def['id'] = cdf['id']
+        custom_field_def['data_type'] = cdf['data_type']
+        custom_field_def['entity_type'] = cdf['definition_entity_type']
+        custom_field_def['facet'] = cdf['facet']
+
+        # Add custom field definition records to custom field defintions keyed by name.
+        custom_field_defs[cdf['name']] = custom_field_def
+
+    return custom_field_defs
+
+# Compare current custom field defintion obtained via the API, to the custom field defintion
+# in the CSV file.
+def compare_custom_field_defs(curr_cfd, csv_cfd):
+    if curr_cfd['data_type'] != csv_cfd['data_type']:
+        print(f"Data_types do not agree.  {curr_cfd['data_type']} != CVS {csv_cfd['data_type']}")
+        sys.exit(1)
+    if curr_cfd['entity_type'] != csv_cfd['entity_type']:
+        print(f"Entity_types do not agree.  {curr_cfd['entity_type']} != CVS {csv_cfd['entity_type']}")
+        sys.exit(1)
+
+    return curr_cfd['id']
+
+# Process the CSV input file custom field defintion dictionary.
+# Return a dictionary of custom field defintions with custom field defintion parameters,
+# custom field values, and vulnerability ID.
+def process_input_file(csv_input_file_name):
+    cdfs = {}
+
+    try:
+        with open(csv_input_file_name, newline='', mode='r', encoding='utf-8-sig') as input_file:
+            reader = csv.DictReader(input_file, delimiter=',', skipinitialspace=True) 
+            for row in reader:
+                cdf_name = row['name']
+                cdfs[cdf_name] = row
+
+    except FileNotFoundError:
+        print(f"ERROR: CSV input file, {csv_input_file_name} not found.")
+        sys.exit(1)
+
+    return cdfs
+
+# Create a custom field defintion.
+def create_custom_field_definition(base_url, headers, cfd_dict):
```

Comment:
What if attachment is provided as a custom field definition type ?

Reply:
This code does not do attachments.  I should add a check for it.  It will be added later.

Reply:
Added a check for the "attachment" `data_type`.